### PR TITLE
feat(mcp-server): log the `initialize` handshake

### PR DESCRIPTION
**Pull request by gilgardosh**
## Why

The connector went blind between Aug 18 and Aug 26 — every tool returned its summary line and no
rows. #4295 fixes the cause (rows lived only in `structuredContent`, a field a client may ignore).

This PR addresses the *second* problem that incident exposed: **the server had no idea the client
had changed.**

Establishing that it *was* the client took a rollback test — checking out `e31e8066`, the commit
that demonstrably worked on Aug 18 including a successful tag write, and running it today reproduces
the failure exactly. Same commit, same server, opposite outcome. Confirming it then meant reading
Claude Desktop's own log directory:

| Date | Log file | Event vocabulary | Outcome |
| --- | --- | --- | --- |
| Aug 10 | `main.log` | `Making remote MCP tool call: accounter_get_charges` → `Remote tool call succeeded` | rows arrived |
| Aug 26 | `claude.ai-web.log` | `[MCP] tool_approval_gate {"toolName":"Accounter:accounter_search_charges",…}` | rows missing |

None of that should have required someone's `~/Library/Logs`. `initialize` (`handler.ts`) never read
`request.params` and never logged, so there was no record of the client's name, its version, the
protocol revision it asked for, or the capabilities it declared. The connector already logs every
tool call and every auth failure — the handshake was the one hop with no trace.

## What

Every `initialize` emits one structured line tagged `event: "mcp_initialize"`, joining `tool_call`
as the second selectable event:

| Field | Why |
| --- | --- |
| `clientName`, `clientVersion` | the fields that date a client-side behaviour change |
| `requestedProtocolVersion` / `servedProtocolVersion` | what was asked vs what we answered |
| `protocolVersionMismatch` | the one field worth alerting on |
| `clientCapabilities` | names only, sorted — values are unbounded and caller-supplied |
| `userId`, `correlationId` | matches `tool_call`, so a session joins across both events |

## Three decisions worth reviewing

**Logged from `dispatchMcpRequest`, not from the `case 'initialize'` that builds the response.**
`handleRpcRequest` is the pure, env-free half and takes only the request — it has neither the caller
nor the correlation id to log. Delegating to it afterwards keeps the response built in exactly one
place, so the two cannot drift. The sync `handleMcpBody` path has no production call sites and stays
silent, which conveniently leaves its existing test of the pure response shape untouched (there's a
test asserting that silence).

**`describeInitializeParams` is total.** `params` is `unknown` off the wire and validated only as a
non-null object *or array*, so every field is narrowed there and anything unexpected degrades to
`null`/`[]`. A malformed handshake must still produce a line — a client sending something the server
cannot parse is precisely the event worth seeing, and an exception would lose it. Covered with
`params` as `null`, a string, an array, a number, and a `clientInfo` that is variously a string, an
array, `null`, or has non-string fields.

**Caller-derived fields are spread beneath the canonical ones**, matching the `tool_call` line.
Without it, `clientInfo` would be an authenticated way to attribute a call to a different `userId` —
there's a test for that specifically. Client strings are clipped before reaching the log.

## Deliberately excluded

- **A `labeledTotals` counter keyed by client version.** `/metrics` is unauthenticated while calling
  a tool requires a token — already flagged in the runbook as worth closing — and client identity is
  a fingerprint of the deployment. It belongs in the log rather than on a public endpoint. Worth
  revisiting once `/metrics` is gated.
- **Protocol-version negotiation.** The server keeps answering `2025-06-18` unconditionally; this
  only records what was asked. Changing what the server advertises is a live behavioural change to a
  connector that has just broken once, and should be decided against a logged mismatch rather than a
  guess — which is what `protocolVersionMismatch` now provides.

## Verification

- 785 tests pass (53 files), 14 of them new; typecheck, eslint and prettier clean
- The `initialize` response is unchanged — asserted directly; this adds a log line, not a new
  response
- Docs: README gains a "Handshake logging" section; runbook gains §3.1 with the field reference and
  `jq` recipes (the cross-reference from README was renumbered to match)

Note the first real run will answer something this investigation could not: which protocol revision
Claude Desktop `1.37937.1` actually negotiates.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/mcp-server/src/mcp/handler.ts [hunk 1]
```diff
@@ -51,6 +51,23 @@ import { listedTools, runSmokeTool, SMOKE_TOOL_NAME } from './tools.js';
...
+ */
+export const MCP_INITIALIZE_EVENT = 'mcp_initialize';
+
+/**
+ * Length cap on a logged client identifier.
+ *
+ * `clientInfo.name` and `clientInfo.version` are caller-supplied strings off the
+ * wire with no schema behind them, so they are clipped before reaching the log.
+ * Mirrors `MAX_MISS_LABEL_LENGTH` in `tools/terminology.ts`, which caps the
+ * other caller-derived text this server records.
```
[the remark this answers is not in the document]
Comment (gilgardosh) [reply packages/mcp-server/src/mcp/handler.ts]: Fair — "mirrors" claimed an alignment that is not there. Reworded in d083655.

The docblock now says what the two actually share (motivation: bounding caller-supplied text) and why the numbers differ on purpose: `MAX_MISS_LABEL_LENGTH` bounds a `/metrics` label, where the constraint is a snapshot that stays readable, while this bounds a log field and only needs to stop an unbounded string. A real identifier — `claude-ai` / `1.37937.1` — is nowhere near either.

File: packages/mcp-server/src/mcp/handler.ts [hunk 2]
```diff
@@ -59,6 +76,64 @@ export const MCP_SERVER_INFO = {
...
+}
+
+/** Clip a caller-supplied identifier for the log, or `null` when absent. */
+function clipClientLabel(value: unknown): string | null {
+  if (typeof value !== 'string' || value.length === 0) {
+    return null;
+  }
+  return value.length <= MAX_CLIENT_LABEL_LENGTH
+    ? value
+    : `${value.slice(0, MAX_CLIENT_LABEL_LENGTH)}\u2026`;
```
[the remark this answers is not in the document]
Comment (gilgardosh) [reply packages/mcp-server/src/mcp/handler.ts]: Correct, and it was a genuine off-by-one: `slice(0, 60)` plus a one-char ellipsis emitted 61. Fixed in d083655 by slicing to `MAX_CLIENT_LABEL_LENGTH - 1`, so the marker counts towards the cap.

The tell was in the test, which asserted `toBeLessThanOrEqual(61)` — a literal that quietly encoded the bug. It now asserts `toHaveLength(MAX_CLIENT_LABEL_LENGTH)` against the exported constant, so raising the cap cannot silently loosen the guarantee, plus a boundary case that a name landing exactly on the cap is returned untouched with no ellipsis.

File: packages/mcp-server/src/mcp/__tests__/handler.test.ts [hunk 3]
```diff
@@ -404,3 +411,186 @@ describe('hasBearerToken', () => {
     expect(hasBearerToken(mockReq('', { authorization: 'Basic xyz' }))).toBe(false);
   });
 });
+
+describe('describeInitializeParams', () => {
+  /**
+   * `params` arrives as `unknown` and is validated only as a non-null object or
+   * array, so every one of these shapes is reachable from the wire. A malformed
+   * handshake still has to produce a line — a client sending something this
+   * server cannot parse is exactly the event worth seeing.
```
[the remark this answers is not in the document]
Comment (gilgardosh) [reply packages/mcp-server/src/mcp/__tests__/handler.test.ts]: You are right, and I checked before changing it: `asJsonRpcRequest` rejects a primitive or null `params` (`jsonrpc.ts:86-88`), so over the wire this really does narrow to object, array, or absent. Reworded in d083655.

The tests still cover those shapes, but the comment now gives the honest reason the helper is total rather than an incorrect one: `params` is typed `unknown`, so the compiler guarantees nothing at this boundary, and the parser's validation is exactly the kind of upstream promise that quietly stops holding — which is the failure mode this whole PR exists to make visible. The dispatch-level test also now says explicitly that it bypasses the parser.
